### ISSUE TITLE
Search API not returning exact matches for some queries

### ISSUE THREAD

**Issue by Kapeli** [rubygems/rubygems.org#972]
The Search API doesn't return exact matches for some queries. As far as I can tell, it happens with queries that have a small length.

For example, if you search for `ib`, the `ib` gem is not part of the search results:

https://rubygems.org/api/v1/search.json?query=ib

**Comment by ETetzlaff** [comment]
I see this bug on production, however, when I pulled down the repo and tried to recreate, I was unable to. 

To try to recreate I just added the ib gem to the Gemfile on the repo and ran the gemcutter task to load gems into the dev db and then plugged in your url.  
http://localhost:3000/api/v1/search.json?query=ib

Results are in the screenshot below.  I'm not sure this is an issue moving forward?

![screen shot 2015-09-05 at 11 08 21 pm](https://cloud.githubusercontent.com/assets/3970630/9702595/0f6ef808-5423-11e5-8d26-4fa8f497df79.png)

**Comment by Kapeli** [comment]
Maybe the production server has some setting that limits the results a search/query can return, which debug doesn't?

**Comment by ETetzlaff** [comment]
It's possible that the production db is doing something, but the query isn't necessarily all that glamorous or complicated so I doubt that a prod db setting is the case here.  After reviewing the logic around it, the only explanation I could have is this (assuming the exact same logic is implemented in prod).

The query logic has a base clause that returns 'ib' in this case, however, there is another stipulation.  The version of that record HAS to be indexed.  If for whatever reason when the gem was created in the prod environment is not indexed, it will not return.  I'm not 100% on the logic surrounding gem creation yet though.

**Comment by stratigos** [comment]
This does not seem to be an issue due to character length, as the following single-character-named gems all return an exact match when searched against a fresh checkout and fresh db dump (as of today):

`l`, `Q`, `h`, `n`, `j`, `f`, `r`, `-`, `v`,`u`,`g`,`_`, `q`,`a`,`c`,`e`,`b`,`p`,`s`,`w`,`i`, ... etc

While `ib` does return an exact match in localdev, the following two-character-named gems do not produce an exact match for me:

`z1`, `lt`, `wb`, `fx`, `rm`, `ko`

There may be more, but perhaps these examples will help uncover why exact matches are not being returned by /search. I checked this set against [production](https://www.rubygems.org/search?utf8=✓&query=ko), and none of these produce an exact match in prod. I tested a few other two-char searches that I know exist in prod, and they produced the expected exact matches.

**Comment by arthurnn** [comment]
we are adding ES to improve the our search, https://github.com/rubygems/rubygems.org/pull/1085 , Should fix this once that is done.

**Comment by vipulnsward** [comment]
Looks like still an issue after #1085

**Comment by arthurnn** [comment]
interesting, that the webui returns the right results: https://rubygems.org/search?utf8=%E2%9C%93&query=ib

**Comment by indirect** [comment]
@sonalkr132 do you think you could look into this?

**Comment by sonalkr132** [comment]
Sure :+1:

**Comment by sonalkr132** [comment]
#1349 was reverted because of #1395

**Comment by sonalkr132** [comment]
Hi, we deployed elasticsearch to the search api. https://rubygems.org/api/v1/search.json?query=ib has `ib` at the top now.

Please let us know if this is still an issue.
